### PR TITLE
Fault Tolerant Manual: Add _rotate_worker_indices utility

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Find the complete PoC here: https://github.com/PyTorchLightning/pytorch-lightning/pull/10568

Parts of #10431

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -570,3 +563,13 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
 
     else:
         raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+
+def _rotate_worker_indices(state, latest_worker_id: int, num_workers: int):
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:568]: Done !

File: pytorch_lightning/utilities/auto_restart.py [hunk 2]
```diff
@@ -570,3 +563,13 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
...
+    """This function is used to rotate the worker indices based on the `latest_worker_id` the training failed
+    on."""
+    if num_workers == 0:
+        return state
+    next_worker_id = latest_worker_id + 1
+    old_to_new_worker_id_map = [((next_worker_id + i) % num_workers, i) for i in range(num_workers)]
+    return {new_id: state[old_id] for old_id, new_id in old_to_new_worker_id_map if old_id in state}
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:575]: Yes, it should be. It is why this is rotating them. Lightning don't support if the state haven't been populated with all workers.

File: pytorch_lightning/utilities/auto_restart.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py:568]: the type (dict) for the state parameter is also missing :)

File: pytorch_lightning/utilities/auto_restart.py [hunk 3]
```diff
@@ -570,3 +563,13 @@ def reload_dataloader_state_dict(dataloader: DataLoader, state_dict: Dict[str, A
 
     else:
         raise MisconfigurationException("This shouldn't happen. Please, open an issue on PyTorch Lightning Github.")
+
+
+def _rotate_worker_indices(state, latest_worker_id: int, num_workers: int):
+    """This function is used to rotate the worker indices based on the `latest_worker_id` the training failed
+    on."""
+    if num_workers == 0:
+        return state
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:572]: let's also handle the error when `latest_worker_id > num_workers - 1` ?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:568]: Done 1

File: pytorch_lightning/utilities/auto_restart.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:572]: Done !

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:
